### PR TITLE
Pass name of BuildRun to Follower

### PR DESCRIPTION
**Pull request by SaschaSchwarze0**
# Changes

When using `shp build run <BUILD_NAME> -F`, then there is no buildrun name available as it gets generated. This caused the following warning:

```
error getting buildrun "" for pod "simple-lib-consumer-d7vm4-chpvd-pod": resource name may not be empty
```

Changing the code to pass the name of the generated build run to the follower.

Also correcting an incorrect message and switching from `'%s'` to `%q` for consistency reasons.

If somebody has a suggestion on what to add to the tests, let me know.

# Submitter Checklist

- [ ] Includes tests if functionality changed/was added
- [ ] Includes docs if changes are user-facing
- [x] [Set a kind label on this PR](https://prow.k8s.io/command-help#kind)  
- [x] Release notes block has been filled in, or marked NONE

# Release Notes

```release-note
Fixed the `shp build run <BUILD_NAME> -F` command to properly pass the name of the generated BuildRun to the log following code to prevent the mis-leading warning about an empty resource name
```

### CODE REVIEW

**Review by otaviof — CHANGES_REQUESTED** [review]
> When using `shp build run <BUILD_NAME> -F`, then there is no buildrun name available as it gets generated. This caused the following warning:
> 
> ```
> error getting buildrun "" for pod "simple-lib-consumer-d7vm4-chpvd-pod": resource name may not be empty
> ```

Let me focus on this part of the PR, please.

Given the problem statement, it seems we have a deeper underlying issue with the CLI. The sub-command `shp build run` has the goal of:

https://github.com/shipwright-io/cli/blob/b42be6944302d384e4fb0b24cd06f3bbcca64a4f/pkg/shp/cmd/build/run.go#L33-L34

Thus, it relies on the [`GenerateName` routine](https://github.com/shipwright-io/cli/blob/b42be6944302d384e4fb0b24cd06f3bbcca64a4f/pkg/shp/cmd/build/run.go#L76-L82) to create the name before the Follower instance is created.

I think we need to investigate if there is a racing condition, like @gabemontero tackled before, and/or we need to improve the error messages.

**Comment by SaschaSchwarze0** [comment]
> > When using `shp build run <BUILD_NAME> -F`, then there is no buildrun name available as it gets generated. This caused the following warning:
> > ```
> > error getting buildrun "" for pod "simple-lib-consumer-d7vm4-chpvd-pod": resource name may not be empty
> > ```
> 
> Let me focus on this part of the PR, please.
> 
> Given the problem statement, it seems we have a deeper underlying issue with the CLI. The sub-command `shp build run` has the goal of:
> 
> https://github.com/shipwright-io/cli/blob/b42be6944302d384e4fb0b24cd06f3bbcca64a4f/pkg/shp/cmd/build/run.go#L33-L34
> 
> Thus, it relies on the [`GenerateName` routine](https://github.com/shipwright-io/cli/blob/b42be6944302d384e4fb0b24cd06f3bbcca64a4f/pkg/shp/cmd/build/run.go#L76-L82) to create the name before the Follower instance is created.
> 
> I think we need to investigate if there is a racing condition, like @gabemontero tackled before, and/or we need to improve the error messages.

Yep, also looking for feedback from @gabemontero. One can also move the instantiation of the Follower to after the BuildRun creation. But I do not know this code in great details, so looking for feedback on what best fits to address this.

**Comment by gabemontero** [comment]
> 

In fact @otaviof @SaschaSchwarze0 the unit test failure is a detected data race `TestStartBuildRunFollowLog` as well with @SaschaSchwarze0 's change here :-) so definitely a race condition of some sort (I don't see the unit test failure being cited explicitly previously in this PR, but apologies if that was inferred somehow), given how I don't remember seeing this during my last set of testing, and the e2e for this command passed previously, as well as this PR (`ok 11 shp build run follow verification`)

The unit tests of course run with race detection, while the e2e's do not.

I'm just now getting tot this at the end of my work day (Wed Jan 12).... apologies for the delay.

I will visit in detail tomorrow AM US Eastern Thur Jan 13 and report back when I have something of note.

**Comment by gabemontero** [comment]
> > 
> 
> In fact @otaviof @SaschaSchwarze0 the unit test failure is a detected data race `TestStartBuildRunFollowLog` as well with @SaschaSchwarze0 's change here :-) so definitely a race condition of some sort (I don't see the unit test failure being cited explicitly previously in this PR, but apologies if that was inferred somehow), given how I don't remember seeing this during my last set of testing, and the e2e for this command passed previously, as well as this PR (`ok 11 shp build run follow verification`)
> 
> The unit tests of course run with race detection, while the e2e's do not.
> 
> I'm just now getting to this at the end of my work day (Wed Jan 12).... apologies for the delay.
> 
> I will visit in detail tomorrow AM US Eastern Thur Jan 13 and report back when I have something of note.

was waiting on a test, and couldn't help myself :-)

so right before I really sign off 
- yeah, the access to Follower.buildRunName only occurs on the various error scenarios (which I why I did not see it in my testing, and the e2e did not see it since the buildrun pod did not have a failure)
- I think the fix that will avoid the data race like we saw in the unit test run https://github.com/shipwright-io/cli/runs/4772823234?check_suite_focus=true, is to get the build run name from the pod label `buildrun.shipwright.io/name` if Follower.buildRunName is empty during the failed Pod block in `onEvent`
- then, in the timeout and non pod event methods, if the field is empty, just adjust the message to not specify the build run name

If you'd like to adjust your PR @SaschaSchwarze0 based on my ideas ^^ go for it.  Otherwise, I don't mind crafting a new PR Thursday my time that you and @otaviof can review, and we sort it out.

Basically, setting Follower struct fields in the Run method of any of the commands is going to eventually be flagged by data race detection.

**Comment by SaschaSchwarze0** [comment]
@gabemontero I tried to change it. Please have a look.

File: pkg/shp/cmd/follower/follow.go [hunk 1]
```diff
@@ -117,6 +117,11 @@ func (f *Follower) stop() {
 
 // OnEvent reacts on pod state changes, to start and stop tailing container logs.
 func (f *Follower) OnEvent(pod *corev1.Pod) error {
+	// the BuildRun name can be empty when the follow was triggered with the buildrun creation
+	if f.buildRunName == "" {
+		f.buildRunName = pod.GetLabels()[buildv1alpha1.LabelBuildRun]
```
Comment (gabemontero) [inline pkg/shp/cmd/follower/follow.go:122]: yeah this setting of buildRunName here will eventually get flagged for a data race either in the unit data race detection or randomly / intermittently with actual usage

OnEvent falls in the same category as Run as I mentioned earlier with respect to multi threadeded 

so change this if block to 

```
brName := f.buildRunName
if len(brName) == 0 {
   brName = pod.GetLabels()[buildv1alpha1.LabelBuildRun]
}
```

And then use `brName` instead of `f.buildRunName` below

File: pkg/shp/cmd/follower/follow.go [hunk 2]
```diff
@@ -130,29 +135,34 @@ func (f *Follower) OnEvent(pod *corev1.Pod) error {
 	case corev1.PodFailed:
 		msg := ""
 		var br *buildv1alpha1.BuildRun
-		err := wait.PollImmediate(1*time.Second, 15*time.Second, func() (done bool, err error) {
-			br, err = f.shpClientset.ShipwrightV1alpha1().BuildRuns(pod.Namespace).Get(f.ctx, f.buildRunName, metav1.GetOptions{})
-			if err != nil {
-				if kerrors.IsNotFound(err) {
+		var err error
+
+		if f.buildRunName != "" {
```
Comment (gabemontero) [inline pkg/shp/cmd/follower/follow.go:140]:
```suggestion
		if brName != "" {
```

File: pkg/shp/cmd/follower/follow.go [hunk 3]
```diff
@@ -130,29 +135,34 @@ func (f *Follower) OnEvent(pod *corev1.Pod) error {
...
+			err = wait.PollImmediate(1*time.Second, 15*time.Second, func() (done bool, err error) {
+				br, err = f.shpClientset.ShipwrightV1alpha1().BuildRuns(pod.Namespace).Get(f.ctx, f.buildRunName, metav1.GetOptions{})
```
Comment (gabemontero) [inline pkg/shp/cmd/follower/follow.go:142]:
```suggestion
				br, err = f.shpClientset.ShipwrightV1alpha1().BuildRuns(pod.Namespace).Get(f.ctx, brName, metav1.GetOptions{})
```

File: pkg/shp/cmd/follower/follow.go [hunk 4]
```diff
@@ -130,29 +135,34 @@ func (f *Follower) OnEvent(pod *corev1.Pod) error {
...
+				if err != nil {
+					if kerrors.IsNotFound(err) {
+						return true, nil
+					}
+					f.Log(fmt.Sprintf("error getting buildrun %q for pod %q: %s\n", f.buildRunName, pod.GetName(), err.Error()))
```
Comment (gabemontero) [inline pkg/shp/cmd/follower/follow.go:147]:
```suggestion
					f.Log(fmt.Sprintf("error getting buildrun %q for pod %q: %s\n", brName, pod.GetName(), err.Error()))
```

File: pkg/shp/cmd/follower/follow.go [hunk 5]
```diff
@@ -130,29 +135,34 @@ func (f *Follower) OnEvent(pod *corev1.Pod) error {
...
+					return false, nil
+				}
+				if br.IsDone() {
 					return true, nil
 				}
-				f.Log(fmt.Sprintf("error getting buildrun %q for pod %q: %s\n", f.buildRunName, pod.GetName(), err.Error()))
 				return false, nil
+			})
+			if err != nil {
+				f.Log(fmt.Sprintf("gave up trying to get a buildrun %q in a terminal state for pod %q, proceeding with pod failure processing", f.buildRunName, pod.GetName()))
```
Comment (gabemontero) [inline pkg/shp/cmd/follower/follow.go:156]:
```suggestion
				f.Log(fmt.Sprintf("gave up trying to get a buildrun %q in a terminal state for pod %q, proceeding with pod failure processing", brName, pod.GetName()))
```

File: pkg/shp/cmd/follower/follow.go [hunk 6]
```diff
@@ -130,29 +135,34 @@ func (f *Follower) OnEvent(pod *corev1.Pod) error {
...
-			}
-			return false, nil
-		})
-		if err != nil {
-			f.Log(fmt.Sprintf("gave up trying to get a buildrun %q in a terminal state for pod %q, proceeding with pod failure processing", f.buildRunName, pod.GetName()))
 		}
+
 		switch {
-		case br == nil:
+		case f.buildRunName != "" && br == nil:
```
Comment (gabemontero) [inline pkg/shp/cmd/follower/follow.go:161]:
```suggestion
		case brName != "" && br == nil:
```

**Review by gabemontero — CHANGES_REQUESTED** [review]
Basically just one needed logical change that had to be applied at multiple lines @SaschaSchwarze0

File: pkg/shp/cmd/follower/follow.go (hunk 1 above)
Comment (SaschaSchwarze0) [reply pkg/shp/cmd/follower/follow.go:122]: @gabemontero the file has other places where it accesses `f.buildRunName` but where there is no Pod object at hand to look it up. So, if that approach is not doable, then we need to revisit when the Follower gets instantiated. In `run.go` should we move this to a place after the buildrun was created and its name is known so that it can be assed into `NewFollower` ?

**Comment by otaviof** [comment]
@SaschaSchwarze0, @gabemontero, I'm still failing to understand what's the real problem we are trying to tackle here. Given the Follower [must generate](https://github.com/shipwright-io/cli/blob/b42be6944302d384e4fb0b24cd06f3bbcca64a4f/pkg/shp/cmd/build/run.go#L33-L34) a new `BuildRun` object.

In this case, when the `BuildRun` creation fails, the CLI must also end in error as well. The `BuildRun` is the starting point that will eventually provide the POD, thus trying to inspect this POD to find out the `BuildRun` name is fundamentally wrong.

**Comment by gabemontero** [comment]
> @SaschaSchwarze0, @gabemontero, I'm still failing to understand what's the real problem we are trying to tackle here. Given the `shp build run` [must generate](https://github.com/shipwright-io/cli/blob/b42be6944302d384e4fb0b24cd06f3bbcca64a4f/pkg/shp/cmd/build/run.go#L33-L34) a new `BuildRun` object.
> 
> When the `BuildRun` creation fails, the CLI must also end up in error as well, and that includes when the `BuildRun` name generated is not available, empty, or any other error. The `BuildRun` is the starting point that will eventually provide the POD, thus trying to inspect this POD to find out the `BuildRun` name is fundamentally wrong.

Yep, let me try to describe the situation here, as well as attempt to build a summary statement, which would most likely should replace the current PR title.  I'll most likely repeat things already understood, but I'm just making sure all the bits are out there.

1) So today, the `Follower` attempts to provide debug information when an error situation occurs.  To do that, there is useful information in the `BuildRun` in addition to the `Pod` object
2) when `shp br logs -F` is the creator of the `Follower`, the `buildRunName` is supplied by the user.  
3) So it can be provided and stored on the `Follower` when it is created
4) when `shp build run -F` is the creator of the `Follower`, the `buildRunName` is generated by the controller
5) taking that `buildRunName` and supplying it to the `Follower` in this case ^^ becomes tricky .... why?
  a)  if you start the `Follower` before supplying the `buildName` to it, you run the chance of data race conditions, both setting and reading a field at the same time
  b) if you set the `buildRunName` on the `Follower` before you start it, but if you do not start it until after the `BuildRun` has been started, meaning you do not start the underlying `PodWatcher` until after the `BuildRun` has started, my experience has been you can miss various `Pod` events, including its completion, if the `BuildRun` / `Pod` fail immediately, and then underlying `PodWatcher` will just hang until its timeout occurs.  It really boils down to how k8s `Watches` work.  By the way, that is why k8s Controllers have relist.  So `PodWatcher` only works for all timing windows if you start it *BEFORE* the `Pod` it wants to watch.   

So ... how to proceed:
1) Tactically, continue to use `BuildRun` information on `Pod` failures or `Pod` never getting created on a best effort basis, where if the label is not available, we still exit but give less information, and the user has to inspect his `Namespace` for the various artifacts.
2) More strategically, fully address 5b), where if the `Pod` watch is created after our request to the controller to create the the `BuildRun` / `Pod`, we don't miss those immediately failures etc.

I would propose we complete 1) with this PR, and open a github issue to tracking 2)

As to that new summary / PR title 

"work around for error diagnostics when the BuildRun name is not available to the Follower / PodWatcher when following logs"

**Comment by otaviof** [comment]
> Yep, let me try to describe the situation here, as well as attempt to build a summary statement, which would most likely should replace the current PR title. I'll most likely repeat things already understood, but I'm just making sure all the bits are out there.
> 
>     1. So today, the `Follower` attempts to provide debug information when an error situation occurs.  To do that, there is useful information in the `BuildRun` in addition to the `Pod` object
> 
>     2. when `shp br logs -F` is the creator of the `Follower`, the `buildRunName` is supplied by the user.
>
>     3. So it can be provided and stored on the `Follower` when it is created

Yes, the `BuildRun` name is provided by the user, and that happens before the actual Follower instance is created.

>     4. when `shp build run -F` is the creator of the `Follower`, the `buildRunName` is generated by the controller

Yes, that's handled by the Controller itself, however, the CLI is fully aware about the generated name, as per:

https://github.com/shipwright-io/cli/blob/b42be6944302d384e4fb0b24cd06f3bbcca64a4f/pkg/shp/cmd/build/run.go#L89-L92

So, could we delay the `NewFollower` after this stage?

>     5. taking that `buildRunName` and supplying it to the `Follower` in this case ^^ becomes tricky .... why?
>        a)  if you start the `Follower` before supplying the `buildName` to it, you run the chance of data race conditions, both setting and reading a field at the same time

Yes, I would in general recommend instantiating those components with all they need in advance.

>        b) if you set the `buildRunName` on the `Follower` before you start it, but if you do not start it until after the `BuildRun` has been started, meaning you do not start the underlying `PodWatcher` until after the `BuildRun` has started, my experience has been you can miss various `Pod` events, including its completion, if the `BuildRun` / `Pod` fail immediately, and then underlying `PodWatcher` will just hang until its timeout occurs.  It really boils down to how k8s `Watches` work.  By the way, that is why k8s Controllers have relist.  So `PodWatcher` only works for all timing windows if you start it _BEFORE_ the `Pod` it wants to watch.

I'm missing the ability of simulating those issues in unit-testing, or maybe end-to-end testing. The changes on this PR are not covered by tests, nor the issues proven by written testing, so we may be introducing fixing changes too quickly.

> So ... how to proceed:
> 
>     1. Tactically, continue to use `BuildRun` information on `Pod` failures or `Pod` never getting created on a best effort basis, where if the label is not available, we still exit but give less information, and the user has to inspect his `Namespace` for the various artifacts.

Wait. We must rely in `Build` and `BuildRun` objects to tell us what are the pods created by the Build Controller. I think we are going too broad on trying to untangle virtually any pod in the cluster.

The `Build` and `BuildRun` are the sources of authority that can lead us to the correct pod, or alternatively if the pod is not yet created, wait for a pod with a proper `ListOptions` to limit the watch.

>     2. More strategically, fully address 5b), where if the `Pod` watch is created after our request to the controller to create the the `BuildRun` / `Pod`, we don't miss those immediately failures etc.

Do Kubernetes watches suppose to work this way? If so, we may need to have a approach where it lists the objects first, and only after start the watch.
 
> I would propose we complete 1) with this PR, and open a github issue to tracking 2)

I think we should prove the problem exists first. The Follower is growing wildly, so we should be more cautions with too much complexity there.

**Comment by SaschaSchwarze0** [comment]
> So, could we delay the `NewFollower` after this stage?

That would actually be my preference.

**Comment by gabemontero** [comment]
> > So, could we delay the `NewFollower` after this stage?
> 
> That would actually be my preference.

We can *IF* we handle the case where the Pod starts and terminates between 
- the creation of the starting of the BuildRun with the generated name 
- the creation of the follower

If we switched that approach in [run.go](https://github.com/shipwright-io/cli/blob/d3e2e62e934fb5befbc040bad732998db640ae94/pkg/shp/cmd/build/run.go#L213-L249) the existing code in [Pod Watcher](https://github.com/shipwright-io/cli/blob/d3e2e62e934fb5befbc040bad732998db640ae94/pkg/shp/reactor/pod_watcher.go) does not handle those 2 bullets, since Pod Watcher is currently on event driven.  To do this switch, we also have to augment Pod Watcher to also include a list, most likely a new item in the select in https://github.com/shipwright-io/cli/blob/d3e2e62e934fb5befbc040bad732998db640ae94/pkg/shp/reactor/pod_watcher.go#L129-L182 that does a list for the Pod we are expecting an event from after X amount of time, where if we get an event, we cancel that timer.

My thought was we not take on that change in this PR.  And again, make a short term and minor tactical change to deal with the name not being available.  But sure, we can always pivot from that.

That is what in my 5b) in my https://github.com/shipwright-io/cli/pull/89#issuecomment-1014562911 was attempting to describe

If what I have written here is still not sufficient for conveying what I am trying to get at, perhaps we look at the code together @otaviof in a buddy programming session or office hours and talk through it.

WRT testing, the current use of data race detection is the testing already in place and has caught the problem is setting the buildRunName field while we reference it (including with this PR).  If we pursued the change I noted ^^ then yes a unit test that captures the timing issue describe here makes sense.

**Comment by otaviof** [comment]
> > > So, could we delay the `NewFollower` after this stage?
> > 
> > 
> > That would actually be my preference.
> 
> [...]
>
> My thought was we not take on that change in this PR. And again, make a short term and minor tactical change to deal with the name not being available. But sure, we can always pivot from that.

I end up going through this process as part of #86, please consider the [latest commit](https://github.com/shipwright-io/cli/pull/86/commits/bf97584d363ed506f8834c8db894800c9e606cd9), and [this comment](https://github.com/shipwright-io/cli/pull/86#issuecomment-1015266859) regarding what do do next.

**Comment by gabemontero** [comment]
OK @otaviof and I talked in our team meeting and landed at:
- for now, allowing the `Follower` to be instantiated in the `RunCommand` `Run` method after the `BuildRun` is generated
- where he currently has this as part of https://github.com/shipwright-io/cli/pull/86
- we agreed the timing window I described about potentially missing Pod events is small, such that we can address that in a separate PR to pod_watcher.go, where he and I reached a high level agreement on how that would look.  I'll either craft a PR for that or open an issue to track after I have a look at his changes, we see how long that takes to complete, etc.

So in the end @SaschaSchwarze0 as you already stated in https://github.com/shipwright-io/cli/pull/86,  if you are willing to wait until https://github.com/shipwright-io/cli/pull/86 lands, and even try out his branch in the meantime, great, we can close this PR out.

Or if you want to adjust this PR where you move the creation of the `Follower` from the `Complete` function to after the `BuildRun` generation in the `Run` function, and pass in the generated name, you can do that too, and that should merge seamlessly into @otaviof 's https://github.com/shipwright-io/cli/pull/86

thanks

**Comment by gabemontero** [comment]
> OK @otaviof and I talked in our team meeting and landed at:
> 
>     * for now, allowing the `Follower` to be instantiated in the `RunCommand` `Run` method after the `BuildRun` is generated
> 
>     * where he currently has this as part of [[WIP][BUILD-373] SHIP-0021: Local Source Upload #86](https://github.com/shipwright-io/cli/pull/86)
> 
>     * we agreed the timing window I described about potentially missing Pod events is small, such that we can address that in a separate PR to pod_watcher.go, where he and I reached a high level agreement on how that would look.  I'll either craft a PR for that or open an issue to track after I have a look at his changes, we see how long that takes to complete, etc.

I have a local commit with the change and unit test for this ^^ ... I'll submit a PR after @otaviof 's #86 merges and I can rebase on top of that 
>

**Comment by gabemontero** [comment]
/close

handled by https://github.com/shipwright-io/cli/pull/86
